## Exploring Two Adwords Campaigns 

Earlier this month, I ran both an Adwords Video (YouTube) and a Google Display campaign. In order to better understand the performance of these campaigns, I want to spend a little bit of time exploring some of the insights that can be gained by studying the placement data. 

In [224]:
import pandas as pd 
import numpy as np 

#import display campaign data and check the basics. 
display=pd.read_csv('imports/display_adwords_placements.csv')
display.info()
display.head(5)

#import video campaign, same deal. 
video=pd.read_csv('imports/Video_adwords_placements.csv')
video.info()
video.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 879 entries, 0 to 878
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Placement      879 non-null    object 
 1   Placement url  879 non-null    object 
 2   Type           879 non-null    object 
 3   Clicks         879 non-null    int64  
 4   Impr.          879 non-null    object 
 5   CTR            879 non-null    object 
 6   Currency code  879 non-null    object 
 7   Avg. CPC       879 non-null    object 
 8   Cost           879 non-null    float64
 9   Conversions    879 non-null    float64
 10  Cost / conv.   879 non-null    float64
 11  Conv. rate     879 non-null    object 
dtypes: float64(3), int64(1), object(8)
memory usage: 82.5+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1474 entries, 0 to 1473
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Placement      1

Placement                                      Placement url  \
0          Blu's Studio  http://youtube.com/channel/UC7m6OSDF6f8We4Kgai...   
1         Triforcefilms  http://youtube.com/channel/UCoDO3V3k1iG-hRGJxL...   
2          GoonCartoons  http://youtube.com/channel/UCKTHJVq9B4aw1enKz1...   
3  Butterfly black subs  http://youtube.com/channel/UCEcLa606RTYxOyuoHO...   
4  Bella e Sofia Macedo  http://youtube.com/channel/UCuZh6cZgT2SMUlSheu...   

              Type    Ad group  Views Currency code Avg. CPV Impr. View rate  \
0  YouTube channel  AdGroup #1      1           USD     0.01     1   100.00%   
1  YouTube channel  AdGroup #1      1           USD     0.01     1   100.00%   
2  YouTube channel  AdGroup #1      1           USD     0.02     1   100.00%   
3  YouTube channel  AdGroup #1      1           USD     0.00     1   100.00%   
4  YouTube channel  AdGroup #1      1           USD     0.01     2    50.00%   

   Cost  Conversions  Cost / conv. Conv. rate  
0  0.01          0.0           0.0      0.00%  
1  0.01          0.0           0.0      0.00%  
2  0.02          0.0           0.0      0.00%  
3  0.00          0.0           0.0      0.00%  
4  0.01          0.0           0.0      0.00%

## Cleaning Up Video Views Data
Taking a look at our data, it seems like we have more than 1,474 placements that only generated one view. It's probably not worth our time to dive too deply on these sources, so let's clean up our data a little bit by getting rid of them. 

In [225]:
better_than_nothing_video_views=video.query('Views >= 2')
better_than_nothing_video_views.head(5)
better_than_nothing_video_views.info()
better_than_nothing_video_views.isna().sum()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 424 entries, 7 to 1472
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Placement      424 non-null    object 
 1   Placement url  424 non-null    object 
 2   Type           424 non-null    object 
 3   Ad group       424 non-null    object 
 4   Views          424 non-null    int64  
 5   Currency code  424 non-null    object 
 6   Avg. CPV       424 non-null    object 
 7   Impr.          424 non-null    object 
 8   View rate      424 non-null    object 
 9   Cost           424 non-null    float64
 10  Conversions    424 non-null    float64
 11  Cost / conv.   424 non-null    float64
 12  Conv. rate     424 non-null    object 
dtypes: float64(3), int64(1), object(9)
memory usage: 46.4+ KB


Placement        0
Placement url    0
Type             0
Ad group         0
Views            0
Currency code    0
Avg. CPV         0
Impr.            0
View rate        0
Cost             0
Conversions      0
Cost / conv.     0
Conv. rate       0
dtype: int64

### No Duplicates, 424 Values 

Excellent. We've already cleaned up our video views significantly. Let's keep going, and do something similar for our display ads data. 

In [226]:
better_than_nothing_display=display.query('Clicks >= 2')
better_than_nothing_display.head(5)
better_than_nothing_display.info()
better_than_nothing_display.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 137 entries, 3 to 878
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Placement      137 non-null    object 
 1   Placement url  137 non-null    object 
 2   Type           137 non-null    object 
 3   Clicks         137 non-null    int64  
 4   Impr.          137 non-null    object 
 5   CTR            137 non-null    object 
 6   Currency code  137 non-null    object 
 7   Avg. CPC       137 non-null    object 
 8   Cost           137 non-null    float64
 9   Conversions    137 non-null    float64
 10  Cost / conv.   137 non-null    float64
 11  Conv. rate     137 non-null    object 
dtypes: float64(3), int64(1), object(8)
memory usage: 13.9+ KB


Placement        0
Placement url    0
Type             0
Clicks           0
Impr.            0
CTR              0
Currency code    0
Avg. CPC         0
Cost             0
Conversions      0
Cost / conv.     0
Conv. rate       0
dtype: int64

## Identifying Viewability Problems 
#### Work In Progress 

In addition to collecting our slice above for further analysis of good opportunities, let's take a moment to collect information that can help us to identify any issues with the viewability of both 

In [227]:
#display.head(5)
no_click_display=display.query('Clicks == 0')
no_click_display_stat=no_click_display.pivot_table(index='Placement', values=['Clicks', 'Impr.'], aggfunc="first")

no_click_display_stat['Impr.']=no_click_display_stat['Impr.'].str.replace(',','')
no_click_display_stat['Impr.']=no_click_display_stat['Impr.'].astype(int)
no_click_with_impressions=no_click_display_stat.query('`Impr.` >= 1')
#no_click_with_impressions.head(20)
click_viewability_issues=pd.merge(no_click_with_impressions,display[['Placement', 'Placement url']], on='Placement', how='left')
click_viewability_issues.to_csv('exports/display_viewability_issues.csv')
print('exported viewability issues to csv')

no_view_video=video.query('Views == 0')
no_view_video_stat=no_view_video.pivot_table(index='Placement', values=['Views','Impr.'], aggfunc="first")
no_view_video_stat['Impr.']=no_view_video_stat['Impr.'].str.replace(',','')
no_view_video_stat['Impr.']=no_view_video_stat['Impr.'].astype(int)
no_view_with_impressions=no_view_video_stat.query('`Impr.` >= 1')
viewability_video_issues=pd.merge(no_view_with_impressions,video[['Placement', 'Placement url']], on='Placement', how='left')
viewability_video_issues.to_csv('exports/video_viewability_issues.csv')
print('exported video viewability issues to csv')

trouble = len(viewability_video_issues) + len(click_viewability_issues)
print("calculating viewability issue count")
print(trouble)

exported viewability issues to csv
exported video viewability issues to csv
calculating viewability issue count
470


## Found 470 Viewability Issues 

We found 1 potential video viewability issue, and 469 potential display viewabilty issues. 

## Looking At High Performing Video Views
Now let's see if we can generate a list of our high performing video view placements. 


In [228]:
#first, let's take a quick look at the stats for views. 
#better_than_nothing_video_views['View rate'].describe()
#ok, let's see if we can look at how these are distributed. 
#better_than_nothing_video_views['View rate'].hist(bins=84)
multiple_views_stat=better_than_nothing_video_views.pivot_table(index='Placement', values=['Views', 'Impr.'], aggfunc="first")
#multiple_views_stat.to_csv("multiple_views.csv")
#multiple_views_rate=multiple_views_stat['Views']/multiple_views_stat['Impr.']
#multiple_views_stat.head(20)
#multiple_views_stat.info()
multiple_views_stat['Impr.']=multiple_views_stat['Impr.'].str.replace(',','')
multiple_views_stat['Impr.']=multiple_views_stat['Impr.'].astype(int)
multiple_views_rate=multiple_views_stat['Views']/multiple_views_stat['Impr.']
multiple_views_stat['vtr']=multiple_views_rate
#cheese=better_than_nothing_video_views.query('Placement in @multiple_views_stat')
#multiple_views_stat['Placement_url']=cheese
merged=pd.merge(multiple_views_stat,better_than_nothing_video_views[['Placement','Placement url']], on='Placement', how='left')
#merged['vtr'].hist(bins=40)
#merged_good_stat=merged.query('.2 < vtr > 1')
#merged_good_stat.head(20)
merged=merged.sort_values(by='Views', ascending=False)
merged.head(20)
merged.to_csv('exports/high_value_view_placements.csv')
print('exported high value view placements!')


exported high value view placements!


### We Got What We Came For! 
We've identified a slice of high value placements. Let's try the same thing with the display data! 

In [229]:
display_stat=better_than_nothing_display.pivot_table(index='Placement', values=['Clicks', 'Impr.'], aggfunc='first')
#display_stat.info()
display_stat['Impr.']=display_stat['Impr.'].str.replace(',','')
display_stat['Impr.']=display_stat['Impr.'].astype(int)
display_stat['ctr']=display_stat['Clicks']/display_stat['Impr.']
#display_stat.head(20)
display_merged=pd.merge(display_stat,better_than_nothing_display[['Placement','Placement url']], on='Placement')
display_merged=display_merged.sort_values(by='Clicks', ascending=False)
display_merged.head(20)
display_merged.to_csv('exports/high_value_display_placements.csv')
print('exported high value display placements!')

exported high value display placements!


# Exported High Value and Problematic Placements 

Now that we've generated our high value video and display placements, we're ready to review and do a little analysis of what we've found. 